### PR TITLE
Add robust_reference function of PREP

### PR DESCRIPTION
**Pull request by Nick3151**
Hi! Aamna and I added some building blocks of pyprep, I added the robust_reference function, which iteratively detects bad channels and interpolates them, until getting a robust estimation of 'true' mean as reference. I also removed the 0.01Hz high pass filtering in Noisydata because I think it is unnecessary(data has already been 1Hz filtered before) and it reports a 'filter too long' warning.
We have not push the main function of PREP here because we'd like to get some feedbacks on the low level functions. We're looking forward to your advice!
@adam2392

### CODE REVIEW

**Comment by adam2392** [comment]
Cool! I’ll check this out later. In the future, you can explicitly request my review so I know you want it.

**Comment by yjmantilla** [comment]
Hi, sorry to interrupt the discussion.

I suspect that sappelhoff put the detrend (High Pass Filter) on a copy inside the Noisydata class constructor because he assumes the signal outside is not detrended (which is done by design on the PREP). This is easier because of his class based implementation of findNoisyChannels.

Putting the detrend outside of Noisydata Class requires that each time you need to deduce the badChannels you:
- copy the original RawMNE object 
- use .filter method for detrend of the copy
- pass it to the Noisydata Class constructor
- then call find_all_bads()

Here I make a copy of the original object because (from what I understood reading the PREP) all the detrends done are temporal, meaning the final signal is not actually detrended.

In https://vislab.github.io/EEG-Clean-Tools/ they say:

> By default, PREP uses a 1 Hz cutoff, but only temporarily filters, so that the final signal is not highpass filtered. This allows you to defer the final choice of high pass cutoff for downstream processing to later."

Basically any time they do a detrend is because a particular procedure needs it, but the detrend itself is not done on the signal that will be final (the one returned by the PREP) but on a copy of it. In particular the signal that enters the referencing stage is not detrended.

In matlab it works easy because the removeTrend is a function (not a method as .filter in RawMNE); the function doesn't alter the original signal, but in MNE the method does, obliguing us to do a copy. So in general it adds more code.

In the robustReference.m code they do a signal = removeTrend(signal) because signal there is a local variable, it is not returned to the outside. So there they don't need to detrend in every call of findNoisyChannels. Actually referenceSignal there is not returned to the outside, what matters is the channels qualified as bad there. The reference signal is computed in performReference.m using the original signal (with the trend) but with the interpolated channels based on the badChannels suggested by robustReference.m .
 
In contrast in performReference.m they use findNoisyChannels(removeTrend(signal)) because signal there is returned to the outside (prepPipeline.m). In our MNE case we will need to use those steps I said.

So in general removing that HP filter adds 2 more line of codes everytime you need find_all_bads(). If the noisy functionality was not in a class but in a more functional-like implementation maybe the overhead would be lower. I would actually put a boolean there indicating robustReference if the signal is detrended or not.

I may be wrong though, but those are my thoughts on the matter. Im not an expert on this. Currently I'm testing a bit of a rough performReference translation I did using the pyprep against some real EEG data processed in the original MATLAB PREP.  

Sorry for long reply and writting, english is not my native language.

**Comment by Nick3151** [comment]
@yjmantilla Thanks for your comment! I agree that it will be more efficient to put high pass filtering in the Noisydata class. However, what I'm concerned about is whether the HP filtering step before findNoisyChannels is necessary. As https://vislab.github.io/EEG-Clean-Tools/ mentioned, the Remove trend(high pass) step is done at the beginning(Step 2 of Processing steps). Since the signal has already been filtered, I suspect we don't need to do filtering again before those downstream steps. 
That is my personal understanding of PREP, you're welcomed to point out any problems you find with it.
By the way, on which EEG datasets are you testing your code? Since our team are also implementing the PREP pipeline in pyprep, maybe we can collaborate on many things

**Comment by yjmantilla** [comment]
Hi @Nick3151 . I've been analyzing those codes of the PREP and I get what you are referring to:

> Processing steps
1.Handle boundary events prior to processing
2.Remove trend (high pass) temporarily to properly compute thresholds
3.Remove line noise without committing to a filtering strategy
4.Robustly reference the signal relative to an estimate of the “true” average reference
5.Detect and interpolate bad channels relative to this reference
6.Produce reports if desired
7.Post process if desired

It does seem to look like the detrend was applied but this is a wrong interpretation (in my view of course), but here are my arguments:

First let stress the "temporarily" quality they put on step 2. This is directly seen in the code of prepPipeline.m which is the actual signal flow: 

```
EEG = input EEG to prepPipeline

% Detrend for line noise removal
EEGNew = removeTrend(EEG)

% Line noise Removal
EEGClean = removeLineNoise(EEGNew)
EEG = EEG - EEGNew + EEGClean

% Notice EEG - EEGNew = EEG with Trend - EEG without Trend = Trend
% EEGClean is EEG without Trend and without Line Noise = Clean EEG Data
% Thus EEG is assigned Trend + Clean EEG Data

% Rereference
EEG = performReference(EEG)
```

Thus the signal that actually enters the performReference code has a Trend. When I started analyzing these codes I was confused by the weird flow path but then I understood it in the light of they not wanting the signal to be actually filtered. Thats why in my previous message I said:

>Basically any time they do a detrend is because a particular procedure needs it, but the detrend itself is not done on the signal that will be final (the one returned by the PREP) but on a copy of it.

The step 2 is done because the Line Noise Removal algorithm needs the signal to be detrended, thats why it appears at the start.

It would be easier to express this in a graphical manner:

prep

Thats why in performReference they use findNoisyChannels(removeTrend(signal)).

I'm testing it on some local datasets from the GRUNECO group of the University of Antioquia (Colombia, UdeA). I will ask the group chief if I can collaborate here, but I'm not really a experienced programmer (still an undergraduate student), but if they say yes I will be happy to.

I want to also note two things that im not sure how important they are:

- Some tresholds sappelhoff use are not exactly the ones in the prep code. For example the robust deviation threshold is 5 in original PREP and sappelhoff uses 3.28. Another one is the thresholds for flat channels which in prep is 10e-10 but sappelhoff uses 1 (?).
- firwin filter in MNE (of scipy) has some differences with firfilt that PREP uses on matlab. I have tested their differences but haven't really arrived at a conclusion:

firs

Firfilt's pop_eegfiltnew (the one applied on the detrend) is described as a 'Hamming windowed sinc FIR filter'. I'm not sure how it compares theoretically with firwin implementation.
 
I have an alpha version of a translation of firfilt to python done with the help of some local students here which was the source of the previous image.

Currently the tests I did regarding the performReference procedure has not return the exact same results as matlab. They agree on some channels but usually the python implementation find more bad channels than matlab. I have tested this with your robustReference traslation as well as a traslation I did. I also tried changing the thresholds but the pattern is still there. Another important thing is that it seems that RANSAC will provide different results randomly (I mean not too different , but like add or take some few channels). This make sense since RANSAC is random-based but for some reason in matlab it always return the same (or maybe I need to do more tests ?)...

EDIT: In line 333 of findNoisyChannels.m I found
```
stream = RandStream('mt19937ar', 'Seed', 435656);
```

So probably thats why matlab's prep is always consistent across runs.

Reedit: Apparently no. Changed that line and it is still consistent across runs.

**Comment by Nick3151** [comment]
Oh, thank you very much! Your interpretation of the workflow is super clear to me! 
Currently I've found the following major differences in Matlab's findNoisyChannels.m and pyprep's Noisydata:

1. Deviation threshold 5 v.s. 3.28 (I see you also found this)
2. Cutoff fraction of bad corr windows 0.01 v.s. 0.1
3. Correlation window for ransac 5s v.s. 4s

Maybe there are some other differences, we can test on different parameters to see if they are important

For the RANSAC issue, I think it's exactly due to what you found, matlab's version sets a random seed so it always return the same, you can also do this in python by setting random.seed()
@yjmantilla

File: test/test_reference.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import pytest
+
+import numpy as np
+import mne
+import random
+from pyprep.reference import robust_reference, remove_reference
+from pyprep.noisy import Noisydata
+
+
+def make_reference_mne_object(sfreq=1000., t_secs=200, n_freq_comps=5,
```
Comment (adam2392) [inline test/test_reference.py:11]: This can be put into a "conftest.py" file, and then imported using pytest.markfixture(). Please read up on how to do this, and then I can show you in person if you're having trouble.

**Comment by adam2392** [comment]
Hi @yjmantilla 

Just wanted to clarify a few things:

1. Are you actively developing something for pyprep? If so, can you just outline what they are? In the spirit of open-source, ideally we don't overlap where necessary.
2. Has @sappelhoff discussed w/ you?

**Review by adam2392 — CHANGES_REQUESTED** [review]
I think overall this is fine. Just do the following:

1. Put code into classes whenever functionality can be modularized that way
2. Make sure you look at scikit-learn style of coding and add a "fit" function to run the entire pipeline.

**Comment by yjmantilla** [comment]
> Hi @yjmantilla
> 
> Just wanted to clarify a few things:
> 
> 1. Are you actively developing something for pyprep? If so, can you just outline what they are? In the spirit of open-source, ideally we don't overlap where necessary.
> 2. Has @sappelhoff discussed w/ you?

Hi, @adam2392 

1) I'm actually trying to translate the whole PREP into python too. I was aware of pyprep so I started using it but I'm not a developer of it per se. As I said I'm just an undergraduate EE student and that was one of the tasks assigned to me.

Mostly I have been doing 1:1 codes of the original matlab implementation of the PREP (because I would get lost sometimes when comparing, and some stuff I just plainly didn't understand). 

For example I have:
- snippets of the firfilt functions
- 2 cases of the removeTrend function
- rough translation of performReference
- rough translation of robustReference

That along their tests against matlab results.

I have spoken with my superiors and they were ok with me collaborating here. So I'm open for collaboration taking into that -although I have used github- I have not really done real collaboration in it.

2) I have not discussed with @sappelhoff  .

**Comment by yjmantilla** [comment]
@Nick3151 I suggest that rather than passing a string of the standards montage to the constructor, pass the actual DigMontage object so that the Noisydata class is more general... I put it like that because my montage is not in the standards. So that in the code there is just:

`self.raw_copy.set_montage(montage)`

This will also get rid of the deprecation warning given to the line:
` montage = mne.channels.read_montage(kind=montage_kind,ch_names=self.raw_copy.ch_names)`

**Comment by sappelhoff** [comment]
So am I understanding this correctly that now two independent parties are going to work on pyprep?

Until now I was of the impression that this is a student project, supervised by @adam2392 :-)

Can everybody say what their goals are, with whom they are working, and in which context?

PS: Let me start: I am currently doing my PhD in Berlin (Max Planck Institute for Human Development), working with EEG data in the domain of human decision making ... I have strong ties to MNE-Python (including mne-bids, etc.) and BIDS, both of which I help to develop as far as my time permits. `pyprep` as in this repo was written by me for analyzing some data I collected during my PhD.

**Comment by Nick3151** [comment]
> @Nick3151 I suggest that rather than passing a string of the standards montage to the constructor, pass the actual DigMontage object so that the Noisydata class is more general... I put it like that because my montage is not in the standards. So that in the code there is just:
> 
> `self.raw_copy.set_montage(montage)`
> 
> This will also get rid of the deprecation warning given to the line:
> ` montage = mne.channels.read_montage(kind=montage_kind,ch_names=self.raw_copy.ch_names)`

Okay, that makes sense

**Comment by Nick3151** [comment]
@sappelhoff 
Sure! I'm Liang, a MS student at Johns Hopkins University(Department of Biomedical Engineering), @AamnaLawrence and I are working on a course project under the guidance of @adam2392 , the TA, and @jovo , the instructor. We hope to implement PREP pipeline in python and contribute to pyprep that works alongside MNE by the end of this semester. 
I'm mainly working on robust_reference and perform_reference function. Aamna is working on removeTrend and findNoisyChannels part. We're now busy with integrating all parts into a complete pipeline. We plan to test the functionality of our codes on the Physionet dataset (https://physionet.org/content/eegmmidb/1.0.0/) by comparing the results with EEGLAB PREP

and in addition: with the hope of addressing the issues outlined in https://github.com/sappelhoff/pyprep/issues/5

**Comment by adam2392** [comment]
@sappelhoff and in addition: with the hope of addressing the issues outlined in https://github.com/sappelhoff/pyprep/issues/5

**Comment by yjmantilla** [comment]
@sappelhoff 

Hi, I'm Yorguin, an undergraduate electronic engineering student at the University of Antioquia (Colombia). I'm working for the Neuropsychology and Behaviour group (GRUNECO) of the university. 

Currently there is an effort to migrate our EEGLAB workflow into python. We also want to make our EEG datasets compliant to BIDS-EEG. The task of translating the PREP was basically assigned to (just) me. Personally my long-term goal is to build a data-science platform for the group that will be able to do a more large-scale multi-study analysis of EEG. 

In general we don't have strong ties to MNE-Python because until now most of the work has been done on MATLAB's EEGLAB. But the goal is to migrate to python and we will probably make use of MNE since it is the equivalent (and it offers BIDS-EEG compatibility that I think you did).

I was aware of the existence of pyprep and I've used it along with some other functions of the PREP that I translated. Mostly the removeTrend, robustReference and performReference, and the firfilt filter. I never really contacted you because I wasn't sure that my codes were good enough (optimization, readability, etc) to be united with pyprep itself. Most of the tests I have done has been against the MATLAB PREP results of a local Alzheimer's disease dataset we are making here.

So in short:

- Working for: GRUNECO - University of Antioquia
- Team : just me as an undergraduate
- Goal: Translate our EEGLAB workflow into python which includes the PREP.
- Context-Domain: Currently we are working on Alzheimer's disease EEG data.

**Comment by sappelhoff** [comment]
okay, great - thanks for introducing yourselves @Nick3151 et al. and @yjmantilla! 

I guess it's up to you to heed @adam2392's first point in https://github.com/NeuroDataDesign/pyprep/pull/2#issuecomment-560164932 so that your work does not overlap too much.

**Comment by adam2392** [comment]
> I guess it's up to you to heed @adam2392's first point in [#2 (comment)](https://github.com/NeuroDataDesign/pyprep/pull/2#issuecomment-560164932) so that your work does not overlap too much.

See https://github.com/sappelhoff/pyprep/issues/5 for full description of proposed changes. Tests to ensure that functionality either improves, or stays the same are as above described by @Nick3151 . 

It would be great to work together if you already have an idea of how you might contribute!

**Comment by yjmantilla** [comment]
Hi @adam2392 and thanks!  I think I can contribute in the following ways:

- I have seen the codes you guys have done (@Nick3151  and @AamnaLawrence) and I think they are of a higher level than those of mine so mainly for the codes you are doing I can test them against data of the laboratory and the results of the matlab's PREP and provide feedback.

- Correct me if I'm wrong but it seems no one is working on the Line Noise Removal stage (or maybe someone will work on it in the future?) so I can contribute on that side. I will try to make the code in a similar way that you guys have done (sappelhoff#5 and the conftest.py thing). Now, this will be in my spare time, meaning I cannot develop at the same speed as you guys.

I would like to know what do you think of these.

**Comment by AamnaLawrence** [comment]
> 
> 
> Hi @adam2392 and thanks! I think I can contribute in the following ways:
> 
>     * I have seen the codes you guys have done (@Nick3151  and @AamnaLawrence) and I think they are of a higher level than those of mine so mainly for the codes you are doing I can test them against data of the laboratory and the results of the matlab's PREP and provide feedback.
> 
>     * Correct me if I'm wrong but it seems no one is working on the Line Noise Removal stage (or maybe someone will work on it in the future?) so I can contribute on that side. I will try to make the code in a similar way that you guys have done ([sappelhoff#5](https://github.com/sappelhoff/pyprep/issues/5) and the conftest.py thing). Now, this will be in my spare time, meaning I cannot develop at the same speed as you guys.
> 
> 
> I would like to know what do you think of these.

Hi @yjmantilla! Nice to know about your interest in pyprep. As far as line noise removal is concerned,
PREP uses spectrum fit and Slepian filters. Fortunately, the MNE notch filter settings could be set to exactly replicate clean line noise removal function by setting parameters like the method to spectrum fit, the bandwidth of the multitaper windowing function and the p-value. We compared the results with MATLAB and found no difference in the outputs.

**Comment by yjmantilla** [comment]
@AamnaLawrence  Ahh I see, I didn't know the multitaper methods could be so easily implemented with the notch filter settings of MNE. So basically you have got the whole PREP practically implemented.

I guess that mainly I will be testing your implementations and provide feedback based on that.  Sorry I couldn't really be more helpful.
